src/code/PreRender.c

```diff
@@ -1,42 +1,638 @@
-#include <ultra64.h>
 #include <global.h>
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0E70.s")
+void PreRender_SetValuesSave(PreRenderContext* this, u32 width, u32 height, void* fbuf, void* zbuf, void* cvg) {
+    this->widthSave = width;
+    this->heightSave = height;
+    this->fbufSave = fbuf;
+    this->cvgSave = cvg;
+    this->zbufSave = zbuf;
+    this->ulxSave = 0;
+    this->ulySave = 0;
+    this->lrxSave = width - 1;
+    this->lrySave = height - 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0EA8.s")
+void PreRender_Init(PreRenderContext* this) {
+    bzero(this, sizeof(PreRenderContext));
+    ListAlloc_Init(&this->alloc);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0ED8.s")
+void PreRender_SetValues(PreRenderContext* this, u32 width, u32 height, void* fbuf, void* zbuf) {
+    this->width = width;
+    this->height = height;
+    this->fbuf = fbuf;
+    this->zbuf = zbuf;
+    this->ulx = 0;
+    this->uly = 0;
+    this->lrx = width - 1;
+    this->lry = height - 1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0F08.s")
+void PreRender_Destroy(PreRenderContext* this) {
+    ListAlloc_FreeAll(&this->alloc);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/PreRender/func_800C0F28.s")
+void func_800C0F28(PreRenderContext* this, Gfx** dListp, void* buf, void* bufSave) {
+    Gfx* dList;
```

Comment:
the pattern we use in most places (at least afaik) is to have a dList variable for the whole display list and then dListHead for the front of it that you do ++ on
So dListp would be dList and Gfx* dList; would be dListHead.

Pretty sure this is the scheme we want to go with anyway, curious on others thoughts